src/test/java/com/spotify/dns/RetainingDnsSrvResolverTest.java

```diff
@@ -134,12 +210,26 @@ public void shouldNotRetainPastEndOfRetentionOnEmptyResults() throws Exception {
     assertThat(resolver.resolve(FQDN).isEmpty(), is(true));
   }
 
+  @Test
+  public void shouldNotRetainPastEndOfRetentionOnEmptyResultsAsync() throws Exception {
+    when(delegate.resolveAsync(FQDN))
+        .thenReturn(CompletableFuture.completedFuture(nodes("aresult")))
+        .thenReturn(CompletableFuture.completedFuture(nodes()));
+
+    resolver.resolveAsync(FQDN).toCompletableFuture().get();
+
+    // expire retained entry
+    Thread.sleep(RETENTION_TIME_MILLIS);
```

Comment:
Sleeping in tests is bad practice.

Reply:
Yeah I can see this causing flakiness, it's copied from `shouldNotRetainPastEndOfRetentionOnEmptyResults()`, we will add a card to fix it later.